### PR TITLE
483 Ignore file passed explictly if matched in "ignore file config"

### LINKED ISSUE #483: Ignore file even when passed as CLI argument
Currently sqlfluff ignores files that match patterns in `.sqlfluffignore` but only if a file is not specified by name directly. 

**For example**

.sqlfluffignore:
```
test.sql
```

test.sql:
```
select a, b, c, d, e fromm table_x
```

Result of `sqlfluff lint .` is empty while result of `sqlfluff lint test.sql` is :
```
== [test.sql] FAIL
L:   1 | P:  22 | L012 | Implicit aliasing of column not allowed. Use explicit
                       | `AS` clause.
L:   1 | P:  27 |  PRS | Found unparsable section: ' table_x'
WARNING: Parsing errors found and dialect is set to 'ansi'. Have you configured your dialect?
```

### Why is it a problem?

I'm currently using sqlfluff with `pre-commit`, which passes a list of changed files to the processes so only the changed files are linted/validated etc. Right now I need to separately add an `exclude:` to the pre-commit config as such:

```yaml
exclude: test.sql
```

### Suggested solution

By default I think sqlfluff should always ignore files that match the patterns defined in `.sqlfluffignore`. If a file is passed explicitly I think it should just log a warning that the file was not linted/parsed because it was ignored. A `--force` flag could be passed to ignore the `.sqlfluffignore` file.

### PR DESCRIPTION
**Pull request by dmateusp**
Closes #483 

## New lint flag

`--not-ignore-files` skips ignore file patterns, by default it is false and ignore file patterns are respected

## New behavior for passing paths explicitly

When a path to a file to be linted is explicitly passed we look for ignore files in all directories that are parents of the file, up to the current directory.

If the current directory is not a parent of the file we only look for an ignore file in the direct parent of the file.

### CODE REVIEW

**Comment by pwildenhain** [comment]
Quick FYI regarding your changes for python 3.5: We're going to be dropping support for 3.5 in #482 which should be getting merged soon

**Comment by dmateusp** [comment]
thanks for the heads up @pwildenhain I think I'll revert the python3.5 specific commits once #482 is merged then

File: src/sqlfluff/linter.py
Comment (alanmcruickshank) [inline src/sqlfluff/linter.py:882]: Can we use `linter_logger.warning()` here rather than `print`? I think that will be more future compatible.

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
One small nit on how we raise the warning, but otherwise looks good. Nice tests in particular!

I wonder whether some of the logic from `_get_potential_ignore_file_locations_for_file` is very somewhat similar to the logic in `sqlfuff.config.ConfigLoader.load_config_up_to_path()`. I can't see an easy way of combining them at the moment, so I'm happy with the logic the way you've done it here - but if you're feeling inspired and see a good way of combining them, then that would be even better.

**Comment by dmateusp** [comment]
@pwildenhain fyi I just hit a minor issue with black, where my outdated version wasn't formatting all the files the same way that the `tox` environment did

what do you think about either having a `requirements_dev.txt` which specifies the version of black to use, or adding `pre-commit` ?

File: src/sqlfluff/linter.py
Comment (dmateusp) [reply src/sqlfluff/linter.py]: The reason I'm using `print` instead of `linter_logger.warning()` is that I wanted to give feedback to the user. I believe we are doing the same here: https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/linter.py#L270

By default `linter_logger.warning()` wasn't showing up, should we change that default instead?

**Comment by pwildenhain** [comment]
I think this is a great idea, and I think `pre-commit` makes the most sense since you would probably want to run `flake8`, `flake8-black`, and `doc8` to handle any of the easy checks before commiting. See #496

**Comment by dmateusp** [comment]
@alanmcruickshank can you review commit c34e3c3a8de711409d11e6797114cd3afad992c6 which is re-using the logic from `sqlfuff.config.ConfigLoader.load_config_up_to_path()`?

File: src/sqlfluff/linter.py
Comment (alanmcruickshank) [reply src/sqlfluff/linter.py]: Good catch. I've just addressed that remaining print statement this in #497. [I've also made warnings red by default too, so they're more obvious).

The warnings *should* come through regardless - can you check that it actually the case though? I've tested on my machine and they are coming through.

File: src/sqlfluff/config.py
Comment (alanmcruickshank) [inline src/sqlfluff/config.py:270]: I really like this pattern! Much tidier.

File: src/sqlfluff/config.py
Comment (alanmcruickshank) [inline src/sqlfluff/config.py:287]: Perhaps a nit, but I think it's important.

Can we rename this method? Reading the logic, it doesn't really find configs, it does something more like finding the paths to search for configs in, given some parameters.

`find_config_locations_up_to_path` or `find_config_paths_up_to_path` could be good names. I think the name  and docstring are a bit msileading at the moment.

*Bonus points*: I think the actual way thing method is working right now it's actually more like a generator, so perhaps we should actually call is `iter_config_paths_to_path`, and instead of building up our `config_stack`, we actually just `yield path` at each point in this method. It could be a much neater way of expressing the same logic.

File: src/sqlfluff/config.py
Comment (alanmcruickshank) [inline src/sqlfluff/config.py:284]: I like this pattern, but I still feel a bit weird about the isolation, and having it instantiate the loader. I think I'm just concerned about what happens when people want to start passing configs in from other places.

I think what would make me feel more comfortable, would be if this was instead a `classmethod`, the only thing it depends on is `find_config_up_to_path`, and I think that method based on your existing changes should actually be a `staticmethod` (on top of my other comments below).

That way when we actually want to use this method in `linter.py` we never have to instantiate it, and we can just call `ConfigLoader.find_ignore_config_files()`.

File: src/sqlfluff/linter.py
Comment (alanmcruickshank) [inline src/sqlfluff/linter.py:877]: see comment above about instantiating the `ConfigLoader`. I think it's wierd that we have to here, but given your other changes I think we can just make it a classmethod and then never need to call `get_global()`.

File: src/sqlfluff/linter.py
Comment (alanmcruickshank) [inline src/sqlfluff/linter.py:886]: I *think* I can see what you're doing here, but this code is quite hard to read. I think it's down to the naming of `path_walk_ignore_file`, which maybe needs a clearer name.

If you can't think of a better name, could you add a few comments next to this code to explain how it works to people who might come across it in future?

File: src/sqlfluff/linter.py
Comment (alanmcruickshank) [inline src/sqlfluff/linter.py:936]: See comments about warnings. I think in #492 warnings should also end up being auto-formatted red so I'd love to use the logging framework where possible.

The reason for that is so that we can be more modular when people start importing `sqlfluff` as a module and not just running it through the cli.

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
Thanks for doing the restructure @dmateusp .

I think the encapsulation here is really nice. I think my remaining feedback is about how we deploy that logic now that it's all in the right place.

Good work, and thanks for taking the time to make this excellent! 🙏

File: src/sqlfluff/linter.py
Comment (dmateusp) [reply src/sqlfluff/linter.py]: Tested now and it's working!

File: src/sqlfluff/config.py
Comment (dmateusp) [reply src/sqlfluff/config.py]: So I refactored to use `yield` but I had to make additional changes to get the same behavior.

In the existing logic we were visiting the given path and going up parent by parent. On each visit we would insert the path to the top of the stack. If I had just replaced with yield, we would be returning the paths in a reversed order compared to the previous logic (which matters for config overwriting priorities, I verified and it broke a test).

So I made some changes to start from the common path and go down the paths instead.

File: src/sqlfluff/linter.py
Comment (dmateusp) [reply src/sqlfluff/linter.py]: I added several comments, let me know what you think. Imo the naming is ok now that there are docs

**Comment by dmateusp** [comment]
@alanmcruickshank I gave it another iteration, thanks for the feedback

**Comment by dmateusp** [comment]
@alanmcruickshank ping :) so this PR doesn't get forgotten since we're looking into adding pre-commit to the repo

**Comment by NiallRees** [comment]
@alanmcruickshank

File: src/sqlfluff/cli/commands.py
Comment (NiallRees) [inline src/sqlfluff/cli/commands.py:280]: I find this a little confusing. Does this seem any better to you?

```suggestion
    help=("Perform the operation regardless of .sqlfluffignore configurations"),
```

File: src/sqlfluff/cli/commands.py
Comment (NiallRees) [inline src/sqlfluff/cli/commands.py:278]: Again I don't quite understand this from first glance. What do you think of something more explicit like

```suggestion
    "--disregard-sqlfluffignores",
```

**Review by NiallRees — CHANGES_REQUESTED** [review]
Just two small nit picks from me with regards to naming and a doc line. Happy to approve once resolved though as rest looks great.

**Review by NiallRees — APPROVED** [review]

**Comment by NiallRees** [comment]
Go for it @dmateusp !

**Pull request merged**
